File: src/domain/entities/messaging/MessageContext.ts
export enum ChatType {
  Direct = 'direct',
  Group = 'group',
  Broadcast = 'broadcast',
  Unknown = 'unknown',
}

export function parseChatType(value: string): ChatType {
  switch (value) {
    case ChatType.Direct:
      return ChatType.Direct;
    case ChatType.Group:
      return ChatType.Group;
    case ChatType.Broadcast:
      return ChatType.Broadcast;
    case ChatType.Unknown:
      return ChatType.Unknown;
    default:
      throw new Error(`Unsupported chat type "${value}".`);
  }
}

/**
 * Messaging metadata extracted from WhatsApp-specific addressing details.
 */
export class MessageContext {
  constructor(
    public readonly chatType: ChatType,
    public readonly remoteJid: string,
    public readonly participantId?: string,
    public readonly senderPhone?: string,
  ) {}
}

File: sdks/java/src/main/java/com/digows/whatsappgateway/messaging/MessageContext.java
package com.digows.whatsappgateway.messaging;

import java.util.List;

/**
 * WhatsApp-specific metadata extracted from the transport envelope around a message.
 *
 * @param chatType normalized chat classification
 * @param remoteJid raw WhatsApp chat JID
 * @param participantId raw participant JID when the message belongs to a group context
 * @param senderPhone resolved sender phone number when available
 * @param mentionedJids raw JIDs mentioned by the message
 * @param quotedMessage quoted message metadata when the message is a reply
 * @param editTarget target message when the payload represents an edit lifecycle
 * @param forwarded whether WhatsApp flagged the message as forwarded
 * @param forwardingScore WhatsApp forwarding score when available
 * @param expirationSeconds disappearing message TTL when available
 * @param viewOnce whether the message was delivered inside a view-once envelope
 */
public record MessageContext(
  ChatType chatType,
  String remoteJid,
  String participantId,
  String senderPhone,
  List<String> mentionedJids,
  QuotedMessage quotedMessage,
  MessageReference editTarget,
  boolean forwarded,
  Integer forwardingScore,
  Integer expirationSeconds,
  boolean viewOnce
)
{
}
